File: ios/BoltCardFieldModule.swift
import Foundation
import UIKit

/// Registry mapping React view tags to BoltCreditCardFieldView instances.
@objcMembers
public class BoltCardFieldRegistry: NSObject {
  public static let shared = BoltCardFieldRegistry()

  private var views: [Int: BoltCreditCardFieldView] = [:]
  private let lock = NSLock()

  public func register(tag: Int, view: BoltCreditCardFieldView) {
    lock.lock()
    views[tag] = view
    lock.unlock()
  }

  public func unregister(tag: Int) {
    lock.lock()
    views.removeValue(forKey: tag)
    lock.unlock()
  }

  public func view(forTag tag: Int) -> BoltCreditCardFieldView? {
    lock.lock()
    let v = views[tag]
    lock.unlock()
    return v
  }
}

/// Companion TurboModule for the BoltCreditCardField Fabric component (ADR-1).
///
/// Provides tokenize(viewTag:) which:
/// 1. Looks up the native BoltCreditCardFieldView via the registry
/// 2. Reads raw card data from byte buffers (never String)
/// 3. Builds JSON body from raw bytes (FR-3.4: no String intermediate for CHD)
/// 4. Encrypts and POSTs to the Bolt tokenizer service (/token endpoint)
/// 5. Zeros all card buffers on every exit path
/// 6. Resolves the Promise with JSON-encoded TokenResult
@objc(BoltCardField)
class BoltCardFieldModule: NSObject {

  private static let tokenizationTimeout: TimeInterval = 30

  @objc
  static func moduleName() -> String {
    return "BoltCardField"
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  func tokenize(_ viewTag: NSNumber,
                 publishableKey: String,
                 apiUrl: String,
                 resolve: @escaping (Any) -> Void,
                 reject: @escaping (String, String, NSError) -> Void) {

    DispatchQueue.main.async {
      self.performTokenize(
        viewTag: viewTag.intValue,
        publishableKey: publishableKey,
        apiUrl: apiUrl,
        resolve: resolve,
        reject: reject
      )
    }
  }

  private func performTokenize(
    viewTag: Int,
    publishableKey: String, // Reserved for future use (cert pinning, rate limiting)
    apiUrl: String,
    resolve: @escaping (Any) -> Void,
    reject: @escaping (String, String, NSError) -> Void
  ) {
    guard let cardFieldView = BoltCardFieldRegistry.shared.view(forTag: viewTag) else {
      reject("VIEW_NOT_FOUND",
             "BoltCreditCardFieldView not found for tag \(viewTag)",
             NSError(domain: "BoltCardField", code: 2))
      return
    }

    // Read raw byte buffers — CHD stays as [UInt8], never becomes String
    var panDigits = cardFieldView.getRawPAN()     // [UInt8] — raw digits 0-9
    var cvvDigits = cardFieldView.getRawCVV()     // [UInt8] — raw digits 0-9
    let expiry = cardFieldView.getRawExpiry()     // "YYYY-MM" (not CHD)
    let postalCode = cardFieldView.getRawPostalCode() // String? (not CHD)
    let network = cardFieldView.getCardNetwork()  // "visa" etc. (not CHD)
    let last4 = cardFieldView.getLast4()          // last 4 only (not full CHD)
    let bin = cardFieldView.getBIN()              // first 6 only (not full CHD)

    // Zero the native field buffers immediately
    cardFieldView.zeroAllBuffers()

    // Determine environment from apiUrl
    let environment: String
    if apiUrl.contains("sandbox") { environment = "sandbox" }
    else if apiUrl.contains("staging") { environment = "staging" }
    else { environment = "production" }

    // Tokenize via the Bolt tokenizer service (NaCl-encrypted)
    let tokenizer = BoltTokenizer(environment: environment)
    tokenizer.tokenize(panDigits: panDigits, cvvDigits: cvvDigits) { result in
      // Zero local copies of CHD
      for i in 0..<panDigits.count { panDigits[i] = 0 }
      for i in 0..<cvvDigits.count { cvvDigits[i] = 0 }

      switch result {
      case .failure(let error):
        reject("E_TOKENIZE_FAILED", error.localizedDescription, error as NSError)

      case .success(let tkResult):
        var tokenResult: [String: Any] = [
          "token": tkResult.token,
          "last4": tkResult.last4.isEmpty ? last4 : tkResult.last4,
          "bin": tkResult.bin.isEmpty ? bin : tkResult.bin,
          "network": network,
          "expiration": expiry,
        ]
        if let postal = postalCode, !postal.isEmpty {
          tokenResult["postal_code"] = postal
        }

        do {
          let jsonData = try JSONSerialization.data(withJSONObject: tokenResult)
          let jsonString = String(data: jsonData, encoding: .utf8) ?? "{}"
          resolve(jsonString)
        } catch let err {
          reject("E_SERIALIZE_FAILED", "Failed to serialize token result", err as NSError)
        }
      }
    }
  }
}

File: android/src/main/java/com/boltreactnativesdk/creditcardfield/BoltCardFieldModule.kt
package com.boltreactnativesdk.creditcardfield

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.UiThreadUtil
import com.facebook.react.uimanager.UIManagerHelper
import org.json.JSONObject

/**
 * Companion TurboModule for the BoltCreditCardField Fabric component.
 *
 * Provides tokenize(viewTag, publishableKey, apiUrl) which:
 * 1. Looks up the native BoltCardFieldView by React view tag
 * 2. Reads raw card data from CharArray buffers (never String for PAN/CVV)
 * 3. Derives the tokenizer environment from apiUrl
 * 4. Encrypts card data via BoltTokenizer (NaCl crypto_box)
 * 5. Zeros all card buffers on every exit path (FR-3.4)
 * 6. Resolves the Promise with JSON-encoded TokenResult
 */
class BoltCardFieldModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = NAME

    @ReactMethod
    fun tokenize(viewTag: Double, publishableKey: String /* reserved for future use */, apiUrl: String, promise: Promise) {
        UiThreadUtil.runOnUiThread {
            try {
                val cardData = readCardData(viewTag.toInt())
                if (cardData == null) {
                    promise.reject("E_VIEW_NOT_FOUND", "BoltCardFieldView not found")
                    return@runOnUiThread
                }

                // Run tokenization off main thread
                Thread {
                    performTokenize(cardData, apiUrl, promise)
                }.start()
            } catch (e: Exception) {
                promise.reject("E_TOKENIZE_FAILED", e.message ?: "Unknown error", e)
            }
        }
    }

    /**
     * Snapshot of card data captured on the UI thread.
     * CharArray copies — caller must zero them after use.
     */
    private data class CardData(
        val panDigits: CharArray,
        val panLength: Int,
        val cvvDigits: CharArray,
        val cvvLength: Int,
        val expiry: String,
        val postalCode: String?,
        val network: String,
        val last4: String,
        val bin: String,
        val zeroViewBuffers: () -> Unit
    )

    private fun readCardData(viewTag: Int): CardData? {
        val uiManager = UIManagerHelper.getUIManager(reactApplicationContext, viewTag) ?: return null
        val view = uiManager.resolveView(viewTag)
        val cardFieldView = findCardFieldView(view) ?: return null

        // Copy raw CharArrays before leaving UI thread
        val pan = cardFieldView.panDigits.copyOf(cardFieldView.panLength)
        val cvv = cardFieldView.cvvDigits.copyOf(cardFieldView.cvvLength)

        return CardData(
            panDigits      = pan,
            panLength      = cardFieldView.panLength,
            cvvDigits      = cvv,
            cvvLength      = cardFieldView.cvvLength,
            expiry         = cardFieldView.getFormattedExpiry(),
            postalCode     = cardFieldView.getPostalCode(),
            network        = cardFieldView.cardNetwork.value,
            last4          = cardFieldView.getLast4(),
            bin            = cardFieldView.getBIN(),
            zeroViewBuffers = { cardFieldView.zeroAllBuffers() }
        )
    }

    private fun performTokenize(cardData: CardData, apiUrl: String, promise: Promise) {
        try {
            val environment = BoltTokenizer.environmentFromApiUrl(apiUrl)
            val tokenizer = BoltTokenizer(environment)

            val result = tokenizer.tokenize(
                panDigits = cardData.panDigits,
                panLength = cardData.panLength,
                cvvDigits = cardData.cvvDigits,
                cvvLength = cardData.cvvLength
            )

            // Zero the CharArray copies now that tokenization succeeded
            cardData.panDigits.fill('\u0000')
            cardData.cvvDigits.fill('\u0000')

            // Zero the view buffers on the UI thread
            UiThreadUtil.runOnUiThread { cardData.zeroViewBuffers() }

            val tokenResult = JSONObject().apply {
                put("token",       result.token)
                put("last4",       result.last4.ifEmpty { cardData.last4 })
                put("bin",         result.bin.ifEmpty { cardData.bin })
                put("network",     cardData.network)
                put("expiration",  cardData.expiry)
                if (!cardData.postalCode.isNullOrEmpty()) {
                    put("postal_code", cardData.postalCode)
                }
            }

            promise.resolve(tokenResult.toString())
        } catch (e: Exception) {
            promise.reject("E_TOKENIZE_FAILED", e.message ?: "Unknown error", e)
        } finally {
            // Ensure CharArrays are zeroed even on exception
            cardData.panDigits.fill('\u0000')
            cardData.cvvDigits.fill('\u0000')
            UiThreadUtil.runOnUiThread { cardData.zeroViewBuffers() }
        }
    }

    private fun findCardFieldView(view: android.view.View?): BoltCardFieldView? {
        if (view is BoltCardFieldView) return view
        if (view is android.view.ViewGroup) {
            for (i in 0 until view.childCount) {
                val found = findCardFieldView(view.getChildAt(i))
                if (found != null) return found
            }
        }
        return null
    }

    companion object {
        const val NAME = "BoltCardField"
    }
}
